### PR TITLE
Support build with openssl-1.1.x

### PR DESCRIPTION
**Pull request by takumakume**
This PR will be able to build the ngx_mruby with openssl-1.1.x.

- Support build with openssl-1.1.x (not supported build opt `enable-tlsext` in openssl-1.1 later)
- Some fixes to CI environment with built-in openssl
  - Upgrade OS to ubuntu:bionic to use libssl-1.1
  - Add environment variable `LD`, Required here: https://github.com/mruby/mruby/blob/b7af9c47b3f6ec656f7e2d7ee4c1d11cda667748/mrbgems/mruby-bin-mrbc/mrbgem.rake#L12
  - Same modification to `Dockerfile.travis_emulate`

CI using openssl-1.0.x is gone.
I thought about doing both 1.0 and 1.1, but I thought it would be complicated.

Instead,
I performed a test overwriting this fix with 1.0.2. passed it.
https://github.com/takumakume/ngx_mruby/pull/1

## Pull-Request Check List

- [ ] Add patches into `src/`.
- [ ] Add test into `test/`. Please see about [test docs](https://github.com/matsumotory/ngx_mruby/tree/master/docs/test).
- [ ] Add docs into `docs/` if you change the features such as [build system](https://github.com/matsumotory/ngx_mruby/tree/master/docs/install), [Ruby methods, class](https://github.com/matsumotory/ngx_mruby/tree/master/docs/class_and_method) and [nginx directives](https://github.com/matsumotory/ngx_mruby/tree/master/docs/directives).

### CODE REVIEW

**Comment by matsumotory** [comment]
@takumakume Thank you for your PR. These patches almost look good to me. Could you try to build on Vagrant using [ngx_mruby vagrant environment](https://github.com/matsumotory/ngx_mruby/blob/master/docs/DEVELOPMENT.md)?

**Comment by takumakume** [comment]
@matsumotory 

Thank you for check!
I forgot Vagrant.

I confirmed that the test passed with vagrant.

commit: [dd51be5](https://github.com/matsumotory/ngx_mruby/pull/433/commits/dd51be5ff06a6d0f6b9ba78fa75b5e61b55936b2)

```sh
vagrant@vagrant:~/ngx_mruby$ ./build/nginx/sbin/nginx -V
:
: (snip)
:
Total: 92
   OK: 92
   KO: 0
Crash: 0
 Time: 17.225444 seconds

"[28403] exit worker process from inline code"
"ngx_mruby: STREAM: mruby_stream_exit_worker_code"
ngx_mruby testing ... Done
test.sh ... successful

vagrant@vagrant:~/ngx_mruby$ ./build/nginx/sbin/nginx -V
nginx version: nginx/1.17.4
built by gcc 4.9.3 (Ubuntu 4.9.3-13ubuntu2)
built with OpenSSL 1.1.1  11 Sep 2018
TLS SNI support enabled
configure arguments: --add-module=/home/vagrant/ngx_mruby --add-module=/home/vagrant/ngx_mruby/dependence/ngx_devel_kit --prefix=/home/vagrant/ngx_mruby/build/nginx --with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt='-g -O0' --with-stream --without-stream_access_module
```

**Comment by matsumotory** [comment]
Do we need the new openssl test in vagrant?

**Comment by matsumotory** [comment]
Or ngx_mruby vagrant already supports openssl-1.1.x, right?

**Comment by takumakume** [comment]
> Do we need the new openssl test in vagrant?

As you said.

It may not need for everyone.
I revert it.

> Or ngx_mruby vagrant already supports openssl-1.1.x, right?

The current vagrant image is not supported openssl-1.1.x.
using `1.0.2` .

```
vagrant@vagrant:~/ngx_mruby$ ./build/nginx/sbin/nginx -V
nginx version: nginx/1.17.4
built by gcc 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11)
built with OpenSSL 1.0.2g  1 Mar 2016
TLS SNI support enabled
configure arguments: --add-module=/home/vagrant/ngx_mruby --add-module=/home/vagrant/ngx_mruby/dependence/ngx_devel_kit --prefix=/home/vagrant/ngx_mruby/build/nginx --with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt='-g -O0' --with-stream --without-stream_access_module
```

**Comment by matsumotory** [comment]
If ngx_mruby test supports openssl 1.1.x by default, ngx_mruby vagrant test also should support openssl 1.1.x, or ngx_mruby test supports both openssl 1.0.x and 1.1.x for compatibility.

**Comment by matsumotory** [comment]
@yyamano How do you think?

**Comment by yyamano** [comment]
I don't use openssl with ngx_mruby, so no strong opinions.

* Do we want to support both versions of openssl? Supporting multiple versions may increase maintenance cost.
* We should use same version of openssl on travis and vagrant.

**Comment by takumakume** [comment]
> Do we want to support both versions of openssl? Supporting multiple versions may increase maintenance cost.

I agree.
but, openssl-1.0.2 supports up to 2019-12-31..

I am difficult to judge.

If support openssl-1.0.x, be able to test for openssl-1.0.x using the travis `matrix` .

- Sample code: https://github.com/takumakume/ngx_mruby/compare/supported-openssl-1.1.x...5ff411bf7b1e
- Test result: https://travis-ci.org/takumakume/ngx_mruby/builds/598630853

> We should use same version of openssl on travis and vagrant.

Yes! I understand.

**Comment by matsumotory** [comment]
@yyamano Thank you for your suggestion. I agree with you. So @takumakume , cloud you implement the vagrant configuration which uses openssl 1.1.x for test as Travis CI?

**Comment by takumakume** [comment]
> So @takumakume , cloud you implement the vagrant configuration which uses openssl 1.1.x for test as Travis CI?

Installed same OpenSSL as Travis CI in vagrant:

```
vagrant@vagrant:~/ngx_mruby$ openssl version
OpenSSL 1.1.1d  10 Sep 2019

vagrant@vagrant:~/ngx_mruby$ ./build/nginx/sbin/nginx -V
nginx version: nginx/1.17.4
built by gcc 4.9.3 (Ubuntu 4.9.3-13ubuntu2)
built with OpenSSL 1.1.1d  10 Sep 2019
TLS SNI support enabled
configure arguments: --add-module=/home/vagrant/ngx_mruby --add-module=/home/vagrant/ngx_mruby/dependence/ngx_devel_kit --prefix=/home/vagrant/ngx_mruby/build/nginx --with-debug --with-http_stub_status_module --with-http_ssl_module --with-cc-opt='-g -O0' --with-stream --without-stream_access_module
```

And test passed in vagrant.

commit: [6403fcf](https://github.com/matsumotory/ngx_mruby/pull/433/commits/6403fcf3e9767b4aeef793a8a6212bd2cfa8b0d4)

File: misc/provision.sh [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
 sudo apt-get -y remove nano
 
 sudo update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-6.0 1000
 
+openssl_version=1.1.1
+curl -sfL https://www.openssl.org/source/openssl-${openssl_version}-latest.tar.gz -o openssl-${openssl_version}.tar.gz
+tar -xzf openssl-${openssl_version}.tar.gz
+rm openssl-${openssl_version}.tar.gz
+cd openssl-${openssl_version}*
+./config --prefix=/usr/local --shared zlib -fPIC >> /dev/null 2>&1
```
Comment (matsumotory) [inline misc/provision.sh]: Why do you remove these logs? These logs are important to investigate the error.

Travis limited the number of stdout line, but the local build is no limitation.

Comment (takumakume) [reply misc/provision.sh]: I understand.
Fixed: https://github.com/matsumotory/ngx_mruby/pull/433/commits/e88ed19401796b776d9d2e8e9eb2a8f82dd432cb

**Comment by matsumotory** [comment]
Thank you for your proposal! Merged this PR now.

**Comment by takumakume** [comment]
Thank you for check!!

**Pull request merged**